### DISCUSSION TITLE
Removed yarpdev --subdevice option in various device wrappers

### DISCUSSION THREAD

**Discussion by randaz81** [robotology/yarp#3078] [Policies and Conventions]
Dear all,
in order to continuously improve Yarp, in terms of efficiency and long-term maintainability, we are going to deprecate `yarpdev --subdevice` option, starting from the next transition from yarp 3.9 to yarp 3.10. This may affect some xml scripts and user applications. An alternate solution is provided below.

### The motivation:
The `subdevice` parameter was implemented by wrapper devices to automatically start the wrapper from a single command line and attach it to a device implementing the interface requested by the device. A clear example is constituted by a video grabber which streams data over a yarp port and a specific camera driver managing the hardware. Over the years, this pattern was copied and pasted on several wrappers but it never became an official standard. Some wrappers were not implementing it, other devices were implementing it in different ways, including validity checks or not, depending and the specific device.
In order to provide an official way to perform this operation, avoiding duplication of code, which is obviously prone to errors, this mechanism is now implemented at a different software level, i.e. by a specific new device called `deviceBundler`.

### What is affected:
All new wrapper devices/nws do not implement the --subdevice parameter: this strategy was already chosen to prepare to this transition. In Yarp 3.9 its use is limited to the few devices listed below: 
- frameGrabberCropper
- AnalogWrapper
- AudioPlayerWrapper
- AudioRecorderWrapper
- JoypadControlServer
- MobileBaseVelocityControl_nws_yarp
- Rangefinder2D_nws_yarp

If yarpdev detects the usage of the --device and --subdevice parameters, an error message will be displayed, showing also the correct command line to use, as described below.

### What is not affected:
This parameter is mostly used by yarpdev command line.
Scripts involving yarpinterface should not be affected by this change, since YRI already implements its own attachment mechanism, which is even more powerful because it allows attaching a device to multiple wrappers.
C++ code is also unaffected by this change, unless the user has not explicitly asked for this mechanism by passing the --subdevice parameter to the Polydriver class, a very uncommon case (which was not detected by searching through the organization repos)

### The alternatives:
Generally speaking, the following command line:
``` 
yarpdev --device devicename --subdevice subdevicename [--param1] [--param2] ...
```
is now replaced by:
```
yarpdev --device deviceBundler --wrapper_device  devicename --attached_device subdevicename [--param1] [--param2] ...
```
which explicitly shows how the system works, without hidden layers:
1. First the device devicebundler is opened.
2. `deviceBundler`opens the two devices `devicename `and `subdevicename.` Please note that optional parameters are passed to all involved devices (this was true also with the previous syntax.
3. `deviceBundler` performs the attach.

See the `deviceBundler` documentation for additional infos.

### Additional raccomandations:
yarpdev executable has two major limitations:
- attaching more than two devices together.
- passing a set of parameters to a single device.
And a single advantage:
- if yarp plugins are compiled statically, it allows to instantiate a couple wrapper-subdevice from a single, convenient command line.

Considering this latter use case, which is mostly relevant on embedded systems, we recommended to prefer the usage of `yarprobotinterface` executable over `yarpdev` in all standard use cases.

### Relevant PRs:
- #3079
- https://github.com/robotology/yarp/pull/3011

**Comment by pattacini** [comment]
cc @robotology/everyone

**Comment by PeterBowman** [comment]
I'm not sure if this was intended, but you can still just invoke the subdevice directly, and [`PolyDriver`](https://github.com/robotology/yarp/blob/b9213d6013e9b771295f391d43b71fb557b54190/src/libYARP_dev/src/yarp/dev/PolyDriver.cpp#L172-L201) (called by `yarpdev`) will silently search for an available wrapper device, instantiate the deviceBundler and perform the attach:

```
yarpdev --device subdevicename [--param1] [--param2] ...
```

This should work as long as you specify a wrapper device during CMake configuration, see `EXTRA_CONFIG WRAPPER`:

```cmake
yarp_prepare_plugin(my_device
                    CATEGORY device
                    TYPE MyDevice
                    INCLUDE MyDevice.hpp
                    EXTRA_CONFIG WRAPPER=wrapper_device_nws)
```

Are there any plans to deprecate/remove this functionality, or can we safely keep using it?

**Reply by traversaro** [reply]
fyi @randaz81
